### PR TITLE
Add cute dsl mla decode op

### PR DESCRIPTION
**Pull request by limin2021**
## 📌 Description

Integrate NVIDIA's CuTe DSL MLA (Multi-Head Latent Attention) decode kernels for Blackwell SM100 into FlashInfer, supporting both BF16/FP16 and FP8 dtypes.

  - Add CuTe DSL MLA decode kernel files (mla_helpers.py, mla_decode_fp16.py, mla_decode_fp8.py) and compilation wrapper (mla_decode.py)
  - Accept tensors from PyTorch and use zero-cost cute.make_tensor layout reinterpretation inside kernel __call__, eliminating ~10 us of Python-side .permute() overhead per call
  - Compile with --enable-tvm-ffi for AOT caching via compile_and_cache_cute_dsl_kernel

  ### Test plan

  - python -m pytest tests/attention/test_cute_dsl_mla_decode.py — 18 tests passing (FP16 + FP8, various batch/head/seq_len configs)
  - Standalone run functions in mla_decode_fp16.py and mla_decode_fp8.py pass
  - pre-commit run --all-files passes

### Funtionality:

All 396 configs PASSED with 0 failures.

  Test matrix:
  - dtype: bfloat16, float8_e4m3fn
  - page_size: 32, 64
  - batch_size: 1, 2, 4, 16, 32, 64, 128, 256, 512, 768, 1024
  - seq_len: 1024, 4096, 8192
  - q_len_per_request: 1, 2, 4
  
  | Status | Count |
  |--------|-------|
  | PASSED | 396   |
  | FAILED | 0     |
  | TOTAL  | 396   |

### Performance:

- **GPU**: NVIDIA Blackwell (SM100a)
- **Model config**: DeepSeek-V3 MLA (128 heads, kv_lora_rank=512, qk_nope_head_dim=128, qk_rope_head_dim=64)
- **KV seq_len**: 8192, **page_size**: 32
- **Timing**: CUPTI, CUDA graph enabled, cold L2 cache, 30 iterations median
- **Date**: 2026-03-11

## 1. FP8 fixed-len (is_var_seq=False → persistent)

```
Config                                             trtllm-gen (ms)   cute-dsl (ms)    Speedup
---------------------------------------------------------------------------------------------
B=1, q=1, s=8192, ps=32, fp8                                0.0159          0.0166      0.96x
B=32, q=1, s=8192, ps=32, fp8                               0.0522          0.0513      1.02x
B=64, q=1, s=8192, ps=32, fp8                               0.0771          0.0704      1.10x
B=128, q=1, s=8192, ps=32, fp8                              0.1430          0.1336      1.07x
B=256, q=1, s=8192, ps=32, fp8                              0.2825          0.2681      1.05x
B=1, q=4, s=8192, ps=32, fp8                                0.0192          0.0185      1.04x
B=32, q=4, s=8192, ps=32, fp8                               0.1307          0.1214      1.08x
B=64, q=4, s=8192, ps=32, fp8                               0.2612          0.2441      1.07x
B=128, q=4, s=8192, ps=32, fp8                              0.4840          0.4533      1.07x
B=256, q=4, s=8192, ps=32, fp8                              0.9927          0.9359      1.06x
```

## 2. FP8 var-seqlen (is_var_seq=True → non-persistent)

```
Config                                             trtllm-gen (ms)   cute-dsl (ms)    Speedup
---------------------------------------------------------------------------------------------
B=1, q=1, s=8192, ps=32, fp8                                0.0159          0.0164      0.97x
B=32, q=1, s=8192, ps=32, fp8                               0.0463          0.0468      0.99x
B=64, q=1, s=8192, ps=32, fp8                               0.0704          0.0640      1.10x
B=128, q=1, s=8192, ps=32, fp8                              0.1264          0.1020      1.24x
B=256, q=1, s=8192, ps=32, fp8                              0.1873          0.1698      1.10x
B=1, q=4, s=8192, ps=32, fp8                                0.0192          0.0184      1.05x
B=32, q=4, s=8192, ps=32, fp8                               0.1181          0.1037      1.14x
B=64, q=4, s=8192, ps=32, fp8                               0.1851          0.1637      1.13x
B=128, q=4, s=8192, ps=32, fp8                              0.3040          0.2930      1.04x
B=256, q=4, s=8192, ps=32, fp8                              0.5964          0.6038      0.99x
```

## 3. BF16 fixed-len (is_var_seq=False → persistent)

```
Config                                             trtllm-gen (ms)   cute-dsl (ms)    Speedup
---------------------------------------------------------------------------------------------
B=1, q=1, s=8192, ps=32, bf16                               0.0241          0.0185      1.30x
B=32, q=1, s=8192, ps=32, bf16                              0.0824          0.0844      0.98x
B=64, q=1, s=8192, ps=32, bf16                              0.1351          0.1283      1.05x
B=128, q=1, s=8192, ps=32, bf16                             0.2566          0.2441      1.05x
B=256, q=1, s=8192, ps=32, bf16                             0.5106          0.4971      1.03x
B=1, q=4, s=8192, ps=32, bf16                               0.0227          0.0224      1.02x
B=32, q=4, s=8192, ps=32, bf16                              0.2136          0.2096      1.02x
B=64, q=4, s=8192, ps=32, bf16                              0.4284          0.4347      0.99x
B=128, q=4, s=8192, ps=32, bf16                             0.7891          0.8124      0.97x
B=256, q=4, s=8192, ps=32, bf16                             1.6007          1.7218      0.93x
```

## 4. BF16 var-seqlen (is_var_seq=True → non-persistent)

```
Config                                             trtllm-gen (ms)   cute-dsl (ms)    Speedup
---------------------------------------------------------------------------------------------
B=1, q=1, s=8192, ps=32, bf16                               0.0241          0.0184      1.31x
B=32, q=1, s=8192, ps=32, bf16                              0.0746          0.0764      0.98x
B=64, q=1, s=8192, ps=32, bf16                              0.1210          0.1126      1.07x
B=128, q=1, s=8192, ps=32, bf16                             0.2196          0.1805      1.22x
B=256, q=1, s=8192, ps=32, bf16                             0.3392          0.3090      1.10x
B=1, q=4, s=8192, ps=32, bf16                               0.0220          0.0214      1.03x
B=32, q=4, s=8192, ps=32, bf16                              0.1841          0.1743      1.06x
B=64, q=4, s=8192, ps=32, bf16                              0.2929          0.2961      0.99x
B=128, q=4, s=8192, ps=32, bf16                             0.5050          0.5384      0.94x
B=256, q=4, s=8192, ps=32, bf16                             1.0073          1.0903      0.92x
```

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **New Features**
  * Added "cute-dsl" backend for MLA decoding with FP16/BF16/FP8 support and a PyTorch-compatible MLA decode API.
  * Added MLA static-tile scheduling helpers and deterministic kernel/workspace handling.

* **Public API**
  * Exposed the new decode API and a torch→Cutlass dtype mapping in the public interface.

* **Tests**
  * End-to-end tests for FP16/BF16/FP8, variable sequence lengths, and API path.

* **Benchmarking**
  * Benchmarks updated to select and exercise the new backend.

### CODE REVIEW

**Comment by limin2021** [comment]
@flashinfer-bot run

**Comment by yongwww** [comment]
@limin2021 pls fix the pre-commit error

**Comment by limin2021** [comment]
> @limin2021 pls fix the pre-commit error

Done

File: flashinfer/mla/_core.py [hunk 1]
```diff
@@ -768,6 +773,39 @@ def trtllm_batch_decode_with_kv_cache_mla(
...
+            raise ValueError(
+                "cute-dsl backend (MLA decode kernel) does not support tensor bmm1_scale, "
+                "please pass a float value"
+            )
+        if isinstance(bmm2_scale, torch.Tensor):
+            raise ValueError(
+                "cute-dsl backend (MLA decode kernel) does not support tensor bmm2_scale, "
+                "please pass a float value"
+            )
+
```
Comment (saltyminty) [inline flashinfer/mla/_core.py:839]: Should we be rejecting cases where unsupported arguments aren't passed in, i.e. sinks, sparse_mla_top_k, enable_pdl?

Comment (limin2021) [reply flashinfer/mla/_core.py:839]: done.

File: flashinfer/cute_dsl/__init__.py
Comment (yzh119) [inline flashinfer/cute_dsl/__init__.py:1]: We are moving kernels out of this directory, maybe create a module

`flashinfer/mla/` and move these kernels under `flashinfer/mla/cute_dsl`?

Comment (limin2021) [reply flashinfer/cute_dsl/__init__.py:1]: done.

File: flashinfer/mla/cute_dsl/mla_decode.py [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+        Float32(1.0),  # output_scale placeholder
+        stream_fake,
+        options="--enable-tvm-ffi --opt-level 2",
+    )
+
+    return compiled_kernel
+
+
+# TODO: query[..., :kv_lora_rank], do we need to remove such kind of slice and move the logic to call routine in the kernel file.
+def cute_dsl_mla_decode(
```
Comment (nvpohanh) [inline flashinfer/mla/cute_dsl/mla_decode.py:283]: @limin2021 Could you check if this backend can be integrated as a backend of https://github.com/flashinfer-ai/flashinfer/blob/main/flashinfer/mla.py#L588 `trtllm_batch_decode_with_kv_cache_mla` with `backend = "cute_dsl"`?

Comment (nvpohanh) [reply flashinfer/mla/cute_dsl/mla_decode.py:283]: Ah I see it now. thanks

File: flashinfer/mla/_core.py
Comment (bkryu) [inline flashinfer/mla/_core.py]: If we are adding `backend='cute-dsl'` to `trtllm_batch_decode_with_kv_cache_mla`, we should add it to the docstring.

Comment (bkryu) [inline flashinfer/mla/_core.py]: Do we need to change the docstring here?

Comment (limin2021) [reply flashinfer/mla/_core.py]: Good catch. Let me add it.

Comment (limin2021) [reply flashinfer/mla/_core.py]: Good catch. Let me add it.

Comment (limin2021) [reply flashinfer/mla/_core.py]: done.

Comment (limin2021) [reply flashinfer/mla/_core.py]: done.

**Comment by limin2021** [comment]
> Since we are exposing a new backend to , we should add a parameter dimension to [test_trtllm_gen_mla.py](https://github.com/flashinfer-ai/flashinfer/blob/043bc43206a7e67eaae0f599f16a49c5d70b27e0/tests/attention/test_trtllm_gen_mla.py#L715-L717) for uniform testing
Hi @bkryu , i've added cute-dsl backend in test_trtllm_gen_mla.py in the latest commit, but I have some new findings.
Currently, the reference is fa2 backend. For some configs, cute-dsl can't pass the reference accuracy check, so I switch to do comparsion with torch ref impl, and then it passed.

**Comment by Observer007** [comment]
> > Since we are exposing a new backend to , we should add a parameter dimension to [test_trtllm_gen_mla.py](https://github.com/flashinfer-ai/flashinfer/blob/043bc43206a7e67eaae0f599f16a49c5d70b27e0/tests/attention/test_trtllm_gen_mla.py#L715-L717) for uniform testing
> > Hi @bkryu , i've added cute-dsl backend in test_trtllm_gen_mla.py in the latest commit, but I have some new findings.
> > Currently, the reference is fa2 backend. For some configs, cute-dsl can't pass the reference accuracy check, so I switch to do comparsion with torch ref impl, and then it passed.

Thanks @limin2021 @bkryu I've several questions:
1. the test named `test_trtllm_gen_mla.py`, it seems the backend should be trtllm-gen, why should we add `cute-dsl` backend to it?
2. As @limin2021 asked, the softmax scale in the file is `scale/sqrt(128+64)`, what does the scale mean here? I think the hidden size of mla decode is 512+64, so the softmax_scale IMO is `1/sqrt(512+64)`.
3. Should we use torch reference as baseline, I believe torch is more solid for reference checking.

**Comment by limin2021** [comment]
> Can you add microbenchmark support for the newly added cute-dsl backend at [benchmarks/routine/attention.py](https://github.com/flashinfer-ai/flashinfer/blob/main/benchmarks/routines/attention.py#L2012-L2028)? We may then want to modify [benchmarks/README.md](https://github.com/flashinfer-ai/flashinfer/blob/main/benchmarks/README.md) about the support. Both should be easily doable with an LLM agent.
Hi @bkryu, this is done in the latest commit.

File: tests/attention/test_trtllm_gen_mla.py [hunk 3]
```diff
@@ -241,6 +286,17 @@ def trtllm_batch_decode_mla(
     if backend == "trtllm-gen":
         if compute_capability[0] != 10:
             pytest.skip("TRTLLM-GEN MLA only supports SM100 and SM103 GPUs")
+    if backend == "cute-dsl":
+        if compute_capability[0] < 10:
+            pytest.skip("cute-dsl MLA requires SM100+")
```
Comment (bkryu) [inline tests/attention/test_trtllm_gen_mla.py]: @limin2021  now most tests seem to be passing with the latest updates to CuTe DSL.

Only failure I am noticing is that we are failing this test on SM120 because the CuTe DSL kernel you added is for SM100f. Can you add a skip here?

Comment (limin2021) [reply tests/attention/test_trtllm_gen_mla.py]: done.

**Review by bkryu — APPROVED** [review]
